### PR TITLE
Decouple Paging and Sorting repositories from `CrudRepository`

### LINKED ISSUE #2537: Split PagingAndSortingRepository and make it a fragment interface
See #2535

### PR DESCRIPTION
**Pull request by schauder**
Closes #2537

### CODE REVIEW

File: src/main/java/org/springframework/data/repository/support/DefaultRepositoryInvokerFactory.java [hunk 1]
```diff
@@ -93,10 +93,7 @@ private RepositoryInvoker prepareInvokers(Class<?> domainType) {
 	@SuppressWarnings("unchecked")
 	protected RepositoryInvoker createInvoker(RepositoryInformation information, Object repository) {
 
-		if (repository instanceof PagingAndSortingRepository) {
```
Comment (mp911de) [inline src/main/java/org/springframework/data/repository/support/DefaultRepositoryInvokerFactory.java:96]: Don't we need the `PagingAndSortingRepositoryInvoker` anymore?

File: src/main/java/org/springframework/data/repository/PagingAndSortingRepository.java [hunk 2]
```diff
@@ -29,7 +29,7 @@
  * @see Page
  */
 @NoRepositoryBean
-public interface PagingAndSortingRepository<T, ID> extends CrudRepository<T, ID> {
+public interface PagingAndSortingRepository<T, ID> extends Repository<T, ID> {
```
Comment (mp911de) [inline src/main/java/org/springframework/data/repository/PagingAndSortingRepository.java:32]: We need similar changes for the reactive repositories.

**Review by mp911de — CHANGES_REQUESTED** [review]

File: src/main/java/org/springframework/data/repository/support/DefaultRepositoryInvokerFactory.java (hunk 1 above)
Comment (schauder) [reply src/main/java/org/springframework/data/repository/support/DefaultRepositoryInvokerFactory.java:96]: That is the main reason why I wanted you to take look at this.

It is not technically necessary, but I guess this comes with a penalty hit. 
No idea how big and relevant that is.

The alternative would be to provide Invokers for different combinations of `CrudRepositories` and `PagingAndSortingRepository` and possibly others.
Smells of code explosion.

File: src/main/java/org/springframework/data/repository/PagingAndSortingRepository.java (hunk 2 above)
Comment (schauder) [reply src/main/java/org/springframework/data/repository/PagingAndSortingRepository.java:32]: Yes, that is understood. 
I'll take care of that once I feel confident I'm heading in the right direction with this.

File: src/main/java/org/springframework/data/repository/PagingAndSortingRepository.java [hunk 3]
```diff
@@ -20,16 +20,19 @@
 import org.springframework.data.domain.Sort;
 
 /**
- * Extension of {@link CrudRepository} to provide additional methods to retrieve entities using the pagination and
- * sorting abstraction.
+ * Repository interface to provide methods to retrieve entities using the pagination and sorting abstraction. In many
```
Comment (mp911de) [inline src/main/java/org/springframework/data/repository/PagingAndSortingRepository.java:23]: `Repository fragment to provide …`

File: src/main/java/org/springframework/data/repository/reactive/ReactiveSortingRepository.java [hunk 4]
```diff
@@ -20,20 +20,24 @@
 
 import org.springframework.data.domain.Sort;
 import org.springframework.data.repository.NoRepositoryBean;
+import org.springframework.data.repository.Repository;
 
 /**
- * Extension of {@link ReactiveCrudRepository} to provide additional methods to retrieve entities using the sorting
- * abstraction.
+ * Repository interface that provides methods to retrieve entities using the sorting abstraction. In many cases it
```
Comment (mp911de) [inline src/main/java/org/springframework/data/repository/reactive/ReactiveSortingRepository.java:26]: `Repository fragment to provide …`

File: src/main/java/org/springframework/data/repository/reactive/RxJava3SortingRepository.java [hunk 5]
```diff
@@ -19,18 +19,21 @@
 
 import org.springframework.data.domain.Sort;
 import org.springframework.data.repository.NoRepositoryBean;
+import org.springframework.data.repository.Repository;
 
 /**
- * Extension of {@link RxJava3CrudRepository} to provide additional methods to retrieve entities using the sorting
- * abstraction.
- *
+ * Repository interface that provides methods to retrieve entities using the sorting abstraction. In many cases this
```
Comment (mp911de) [inline src/main/java/org/springframework/data/repository/reactive/RxJava3SortingRepository.java:25]: `Repository fragment to provide …`

File: src/main/kotlin/org/springframework/data/repository/kotlin/CoroutineSortingRepository.kt [hunk 6]
```diff
@@ -18,25 +18,30 @@ package org.springframework.data.repository.kotlin
 import kotlinx.coroutines.flow.Flow
 import org.springframework.data.domain.Sort
 import org.springframework.data.repository.NoRepositoryBean
+import org.springframework.data.repository.Repository
 
 /**
- * Extension of [CoroutineCrudRepository] to provide additional methods to retrieve entities using the sorting
+ * Repository interface that provides methods to retrieve entities using the sorting
```
Comment (mp911de) [inline src/main/kotlin/org/springframework/data/repository/kotlin/CoroutineSortingRepository.kt:24]: `Repository fragment to provide …`

File: src/main/kotlin/org/springframework/data/repository/kotlin/CoroutineSortingRepository.kt [hunk 7]
```diff
@@ -18,25 +18,30 @@ package org.springframework.data.repository.kotlin
...
 
-	/**
-	 * Returns all entities sorted by the given options.
-	 *
-	 * @param sort must not be null.
-	 * @return all entities sorted by the given options.
-	 * @throws IllegalArgumentException in case the given [Sort] is null.
-	 */
-	fun findAll(sort: Sort): Flow<T>
+    /**
```
Comment (mp911de) [inline src/main/kotlin/org/springframework/data/repository/kotlin/CoroutineSortingRepository.kt:39]: Indentation has changed to spaces here…

**Review by mp911de — CHANGES_REQUESTED** [review]
Do we need to update the documentation as well?

**Comment by mp911de** [comment]
Thanks. I'll handle the merge tomorrow and follow up with modules that require changes.

**Comment by schauder** [comment]
Thanks. Let me know so we can handle the follow up changes together.

**Comment by mp911de** [comment]
That's merged and polished now.

**Pull request closed without merging**